reframe/core/schedulers/slurm.py

```diff
@@ -692,3 +724,316 @@ def _extract_attribute(self, attr_name, node_descr, sep=None):
+
+        for dirpath, dirnames, files in firecrest_walk(job._remotedir):
+            local_dirpath = join_and_normalize(
+                job._localdir,
+                os.path.relpath(
+                    dirpath,
+                    job._remotedir
+                )
+            )
+            for d in dirnames:
+                new_dir = join_and_normalize(local_dirpath, d)
+                self.log(f'Creating local directory {new_dir}')
+                os.makedirs(new_dir)
+
+            for (f, modtime, fsize) in files:
+                norm_path = join_and_normalize(dirpath, f)
+                local_file_path = join_and_normalize(local_dirpath, f)
+                if self._remote_filetimestamps.get(norm_path) != modtime:
+                    self.log(f'Downloading file {f} in {local_dirpath}')
+                    self._remote_filetimestamps[norm_path] = modtime
+                    _download(
+                        norm_path,
+                        local_file_path,
+                        fsize
+                    )
+
+                new_modtime = os.path.getmtime(local_file_path)
+                self._local_filetimestamps[local_file_path] = new_modtime
+
+    def submit(self, job):
+        job._localdir = os.getcwd()
+        job._remotedir = os.path.join(
+            self._remotedir_prefix,
+            os.path.relpath(os.getcwd(), job._stage_prefix)
+        )
+
+        if job._clean_up_stage:
+            # Create clean stage directory in the remote system
+            try:
+                self.client.simple_delete(self._system_name, job._remotedir)
+            except fc.HeaderException:
+                # The delete request will raise an exception if it doesn't
+                # exist, but it can be ignored
+                pass
+
+        self.client.mkdir(self._system_name, job._remotedir, p=True)
+        self.log(f'Creating remote directory {job._remotedir} in '
+                 f'{self._system_name}')
+
+        self._push_artefacts(job)
+
+        intervals = itertools.cycle([1, 2, 3])
+        while True:
+            try:
+                # Make request for submission
+                submission_result = self.client.submit(
+                    self._system_name,
+                    os.path.join(job._remotedir, job.script_filename),
+                    local_file=False
+                )
+                break
+            except fc.FirecrestException as e:
+                stderr = e.responses[-1].json().get('error', '')
+                error_match = re.search(
+                    rf'({"|".join(self._resubmit_on_errors)})', stderr
+                )
+                if not self._resubmit_on_errors or not error_match:
+                    raise
+
+                t = next(intervals)
+                self.log(
+                    f'encountered a job submission error: '
+                    f'{error_match.group(1)}: will resubmit after {t}s'
+                )
+                time.sleep(t)
```

Comment:
The problem with this implementation is that it is blocking and the `submit()` call must non-blocking because otherwise no other job will be able to proceed. Waiting for the artefacts to be uploaded can take quite some time. Also, sleeping in intervals of 1-3s waiting for FC to accept the submission, I think it's too coarse if this were to be done inside the `submit()`.

What if you spawn other Python processes to handle the push/submit/pull and from the backend you simply control them, like the SSH backend does. There are two ways to do this:

1. The more Pythonic: extend our `_ProcFuture` interface to handle `multiprocess.Process` object. Essentially, make a base future with the common functionality and have the current `_ProcFuture` and the `_MultiprocessProcFuture` expand from this.
2. The simplest (maybe): Create a separate simple special purpose python script that would handle the Firecrest connection (push/submit/pull) which you control from the backend in a non-blocking fashion.

Reply:
> What if you spawn other Python processes to handle the push/submit/pull and from the backend you simply control them, like the SSH backend does.

Yes, this is definitely an important improvement. I skipped it to have an implementation sooner, but you are right. Also there is a whole async version of pyfirecrest that would be a much better solution but I was kinda waiting to drop python 3.6 support in reframe 😅 I guess pyfirecrest requires 3.7 anyway though, so I can directly use it (?) 🤔